rllab/envs/dmcontrol_env.py

```diff
@@ -0,0 +1,130 @@
+import pygame
+import numpy as np
+from PIL import Image
+from dm_control import suite
+from dm_control.rl.environment import StepType
+from dm_control.suite.wrappers import pixels
+from rllab.envs.base import Env, Step
+from rllab.core.serializable import Serializable
+from rllab.spaces.box import Box
+from rllab.spaces.discrete import Discrete
+
+
+'''
+This environment will use dm_control toolkit to train and simulate your model.
+Domain name and task name used for dm_control is showed below: (domain name/task name)
+
+ball_in_cup/catch
+cartpole/swingup_sparse
+cartpole/balance_sparse
+cartpole/swingup
+cartpole/balance
+cheetah/run
+finger/turn_hard
+finger/spin
+finger/turn_easy
+fish/upright
+fish/swim
+hopper/stand
+hopper/hop
+humanoid/stand
+humanoid/run
+humanoid/walk
+manipulator/bring_ball
+pendulum/swingup
+point_mass/easy
+reacher/hard
+reacher/easy
+swimmer/swimmer6
+swimmer/swimmer15
+walker/stand
+walker/run
+walker/walk
+'''
+
+
+def flat_shape(observation):
+    dim = 0
+    for key, value in observation.items():
+        dim += int(np.prod(value.shape))
+    return dim
+
+def flat_observation(observation):
+    new_observation=[]
+    for key, value in observation.items():
+        new_observation += value.flatten().tolist()
+    observation = np.array(new_observation)
+    return observation
+
+
+class DMControlEnv(Env, Serializable):
+    def __init__(self, domain_name, task_name = "swingup", 
+            display = True, width = 320, height = 240, pixel_key = 'rbg', pixels_only = False, *arg, **kwargs):
+        Serializable.quick_init(self, locals())
+        pygame.init()
+
+        #create pygame window
+        self.display_window = pygame.display.set_mode((width,height))
+
+        #create dm control env
+        self.env = suite.load(domain_name=domain_name, task_name=task_name)
+        self.wrapped = pixels.Wrapper(self.env,
+                      observation_key=pixel_key,
+                      pixels_only=pixels_only,
+                      render_kwargs={'width': width, 'height': height})
+
+        self.pixel_key = pixel_key
+        self._display = display
+        self._observation_spec = self.wrapped.observation_spec()
+        self._action_spec = self.wrapped.action_spec()
+        self._total_reward=0
+
+    def step(self, action):
+        time_step = self.wrapped.step(action)
+        if self.display == True:
+            print (time_step)
+            pixels_img=time_step.observation[self.pixel_key]
+            self._set_image(pixels_img)
+        self._total_reward += time_step.reward
+        
+        return flat_observation(time_step.observation), time_step.reward, \
```

Comment:
This needs to return the `rllab.envs.base.Step` data type.

Reply:
I have followed the return type according to the base.py.
However, reward should be a float value, since it will multiply a scale in normalized_env.py. 
Observation should be a matrix or a array,  since it will be normalized in normallized_env.py.

-rllab/envs/base.py
>(observation, reward, done, info)
>observation : agent's observation of the current environment
>reward [Float] : amount of reward due to the previous action
>done : a boolean, indicating whether the episode has ended
>info : a dictionary containing other diagnostic information from the previous action

-rllab/envs/normalized_env.py

    @overrides
    def step(self, action):
        if isinstance(self._wrapped_env.action_space, Box):
        # rescale the action
            lb, ub = self._wrapped_env.action_space.bounds
            scaled_action = lb + (action + 1.) * 0.5 * (ub - lb)
            scaled_action = np.clip(scaled_action, lb, ub)
        else:
            scaled_action = action
            wrapped_step = self._wrapped_env.step(scaled_action)
            next_obs, reward, done, info = wrapped_step
        if self._normalize_obs:
            next_obs = self._apply_normalize_obs(next_obs)
        if self._normalize_reward:
            reward = self._apply_normalize_reward(reward)
        return Step(next_obs, reward * self._scale_reward, done, **info)

Reply:
Example from gym_env: https://github.com/ryanjulian/rllab/blob/integration/rllab/envs/gym_env.py#L114

You are correct about the member types. I'm just saying that it should return Step(...)